File: frontend/lib/shell.ts
"use client";

/**
 * shell — "Native Shell + Web Work Area" архитектурын web талын гэрээ.
 *
 * Native бүрхүүл (одоо Swift/AppKit, дараа Tauri) нь нэвтрэлт, толгой хэсэг,
 * цэс, хөл, төхөөрөмжийн хандалтыг өөртөө авдаг. Тэр үед web app нь өөрийн
 * chrome-оо нуугаад зөвхөн ажлын муж болж рендерлэгдэнэ. Хөтөч дээр shell
 * байхгүй тул энэ файлын бүх зүйл `null`/`false` буцаана — өөрөөр хэлбэл
 * хөтчийн горим ямар ч нөхцөлд өөрчлөгдөхгүй.
 *
 * Гэрээний бүрэн бүртгэл: `docs/SHELL_CONTRACT.md`.
 */

import { useSyncExternalStore } from "react";

export type ShellPlatform = "macos" | "windows" | "linux" | "ios" | "android" | "kiosk" | "pos";

export interface GeregeShell {
  /** Гэрээний semver. Одоо "1.0". */
  version: string;
  platform: ShellPlatform;
  /** Тухайн бүрхүүлд ҮНЭХЭЭР хэрэгжсэн чадварууд. Зарлагдаагүй method дуудвал
   *  invoke() reject хийнэ. */
  capabilities: string[];
  invoke<T>(method: string, params?: Record<string, unknown>): Promise<T>;
  /** Буцаах утга нь бүртгэлээ цуцлах функц. */
  on(event: string, handler: (payload: unknown) => void): () => void;
}

declare global {
  interface Window {
    GeregeShell?: GeregeShell;
  }
}

/** Гэрээгээр тогтсон method нэрс. Хаана ч мөрөн дотор бичихгүй — нэр солигдвол
 *  энэ нэг газраас л өөрчлөгдөнө. */
export const SHELL_METHODS = {
  /** Session дуусахад бүрхүүлийн нэвтрэлтийн урсгалыг эхлүүлнэ. */
  AUTH_RE_LOGIN: "auth.reLogin",
  NOTIFY_SHOW: "notify.show",
  BADGE_SET: "badge.set",
  BIOMETRIC_AUTHENTICATE: "biometric.authenticate",
  EXTERNAL_OPEN: "external.open",
  PRINT_SYSTEM: "print.system",
  FS_SAVE_AS: "fs.saveAs",
  /** Тенантын цэс өөрчлөгдсөнийг мэдэгдэнэ — бүрхүүл өөрийн native цэсээ дахин
   *  татах боломжтой болно. */
  MENU_CHANGED: "menu.changed",
} as const;

/** Гэрээгээр тогтсон event нэрс. Бүрхүүлээс web рүү чиглэнэ. */
export const SHELL_EVENTS = {
  NAVIGATE: "shell:navigate",
  SEARCH: "shell:search",
  MENU_REFRESH: "shell:menu-refresh",
} as const;

/** Гэрээгээр тогтсон capability нэрс. */
export const SHELL_CAPABILITIES = {
  BIOMETRIC: "biometric",
  PRINT_SYSTEM: "print.system",
  /** `fs.saveAs` method-ын чадвар. Чадвар нь боломж, method нь дуудлага. */
  FS_SAVE: "fs.save",
  NOTIFY: "notify",
  BADGE: "badge",
  EXTERNAL_OPEN: "external.open",
  SECURE_STORE: "secure-store",
  MENU_NATIVE: "menu.native",
} as const;

export interface ShellNavigatePayload {
  path: string;
}

export interface ShellSearchPayload {
  query: string;
}

/**
 * Одоогийн бүрхүүл. SSR-д үргэлж `null` — `window` байхгүй.
 *
 * Бүрхүүл объектыг document start дээр inject хийдэг тул hydration-ы үед аль
 * хэдийн байрандаа байдаг; ижил объектын лавлагаа буцдаг нь
 * `useSyncExternalStore`-ын snapshot-ыг тогтвортой байлгана.
 */
export function getShell(): GeregeShell | null {
  if (typeof window === "undefined") return null;
  return window.GeregeShell ?? null;
}

export function hasCapability(cap: string): boolean {
  return getShell()?.capabilities?.includes(cap) ?? false;
}

// Бүрхүүл ажиллах хугацаандаа солигддоггүй (inject хийгээд л дуусдаг), тул
// захиалга нь юу ч хийхгүй — snapshot нэг л удаа уншигдана.
const subscribeToShell = () => () => {};
const getServerShell = () => null;

/**
 * React хук: бүрхүүл ба түүн дотор ажиллаж байгаа эсэх.
 *
 * Сервер дээр `inShell` нь үргэлж `false` тул анхны markup хөтчийнхтэй адил
 * гарч, hydration-ы дараа л бүрхүүлийн харагдац руу шилжинэ.
 */
export function useShell(): { shell: GeregeShell | null; inShell: boolean } {
  const shell = useSyncExternalStore(subscribeToShell, getShell, getServerShell);
  return { shell, inShell: shell !== null };
}

export type ShellInvokeResult<T> =
  | { ok: true; value: T }
  | { ok: false; reason: "no-shell" | "error" | "timeout"; error?: Error };

/**
 * Хэзээ ч throw хийхгүй, хэзээ ч мөнхөд өлгөөстэй үлдэхгүй invoke.
 *
 * Дуудагч тал ихэвчлэн "бүрхүүл үүнийг барьж авсан уу, үгүй юу" гэдгийг л
 * мэдэх шаардлагатай байдаг — бүрхүүл дэмжихгүй бол, алдвал, эсвэл огт хариу
 * буцаахгүй бол бүгд адилхан "web талын fallback-аа ажиллуул" гэсэн утгатай.
 */
export async function invokeShell<T>(
  method: string,
  params?: Record<string, unknown>,
  timeoutMs = 20000,
): Promise<ShellInvokeResult<T>> {
  const shell = getShell();
  if (!shell) return { ok: false, reason: "no-shell" };

  let timer: ReturnType<typeof setTimeout> | undefined;
  const timeout = new Promise<{ ok: false; reason: "timeout" }>((resolve) => {
    timer = setTimeout(() => resolve({ ok: false, reason: "timeout" }), timeoutMs);
  });

  try {
    const invocation = shell
      .invoke<T>(method, params)
      .then((value) => ({ ok: true, value }) as const)
      .catch((error: unknown) => ({
        ok: false,
        reason: "error",
        error: error instanceof Error ? error : new Error(String(error)),
      }) as const);
    return await Promise.race([invocation, timeout]);
  } finally {
    clearTimeout(timer);
  }
}

File: desktop-tauri/src-tauri/src/shell.rs
//! Bridge Contract v1 — `window.GeregeShell`.
//!
//! Гэрээ нь `docs/SHELL_CONTRACT.md`-д тодорхойлогдсон бөгөөд өөрчлөгдөхгүй
//! спецификац. Энэ модуль түүний Tauri талын хэрэгжилт: inject хийх скрипт,
//! `invoke()`-ийн диспетчер, event илгээх туслахууд.

use serde_json::{json, Value};
use tauri::{AppHandle, Manager, WebviewWindow};

use crate::config::Endpoints;

pub const CONTRACT_VERSION: &str = "1.0";

/// Гэрээгээр тогтсон event нэрс.
pub const EVENT_NAVIGATE: &str = "shell:navigate";
pub const EVENT_SEARCH: &str = "shell:search";
pub const EVENT_MENU_REFRESH: &str = "shell:menu-refresh";

/// Ажлын мужийг агуулах цонхны шошго.
pub const MAIN_WINDOW: &str = "main";

/// Build target-аас гарах платформын нэр. Гэрээний `platform` талбар ба
/// `<html data-shell="...">` атрибутын утга болно.
pub fn platform() -> &'static str {
    if cfg!(target_os = "macos") {
        "macos"
    } else if cfg!(target_os = "windows") {
        "windows"
    } else {
        "linux"
    }
}

/// ЗӨВХӨН энэ бүрхүүлд үнэхээр хэрэгжсэн чадварууд.
///
/// `biometric` энд алга: Tauri-гийн биометр залгуур нь мобайл платформынх,
/// desktop дээр хэрэгжилт байхгүй. `secure-store` мөн алга — гэрээний v1-д
/// түүнийг ашиглах method тодорхойлогдоогүй тул зарлах нь утгагүй болно.
pub fn capabilities() -> Vec<&'static str> {
    vec![
        "notify",
        "badge",
        "external.open",
        "print.system",
        "fs.save",
        "menu.native",
    ]
}

/// JS эх бичвэрт шууд суулгаж болох, бүрэн escape хийгдсэн string literal.
///
/// Native талаас JS руу орох утга бүр үүгээр эсвэл JSON-оор л явна. Гэрээний
/// аюулгүй байдлын шаардлага #3: ямар ч эх сурвалжаас ирсэн текст код болж
/// ажиллах боломжгүй байх ёстой.
pub fn js_string_literal(value: &str) -> String {
    let encoded = match serde_json::to_string(value) {
        Ok(encoded) => encoded,
        // Энгийн мөрийг JSON болгоход алдаа гарах боломж бараг байхгүй ч
        // panic хийхээс хоосон мөр буцаах нь дээр.
        Err(_) => return "\"\"".to_string(),
    };
    // JSON нь U+2028/U+2029-ийг түүхийгээр нь үлдээдэг; JS-ийн зарим парсер
    // эдгээрийг мөр таслал гэж уншина.
    encoded.replace('\u{2028}', "\\u2028").replace('\u{2029}', "\\u2029")
}

/// Document start дээр inject хийх скрипт.
///
/// Гол frame-д л ажиллана: Tauri-гийн initialization script нь дэд frame-д
/// хүрэхгүй бөгөөд доорх `window.top !== window` шалгалт нэмэлт хамгаалалт.
pub fn init_script(endpoints: &Endpoints) -> String {
    let config = json!({
        "version": CONTRACT_VERSION,
        "platform": platform(),
        "capabilities": capabilities(),
        "apiBase": endpoints.api_url,
    });
    let config_json = config.to_string();

    format!(
        r#"
(function () {{
  // Гэрээ зөвхөн ажлын мужид хамаарна. Дэд frame нь native чадваруудад
  // хүрэх шаардлагагүй.
  if (window.top !== window) {{ return; }}
  if (window.GeregeShell) {{ return; }}

  var config = JSON.parse({config});
  var internals = window.__TAURI_INTERNALS__;
  function ipc(command, payload) {{
    if (!internals || typeof internals.invoke !== 'function') {{
      return Promise.reject(new Error('shell: IPC боломжгүй'));
    }}
    return internals.invoke(command, payload);
  }}

  // Native тал бүх хариугаа энэ нэг цэгээр буцаана (гэрээний шаардлага #4).
  window.__geregeShellEmit = function (name, payloadJSON) {{
    var detail = null;
    try {{ detail = payloadJSON ? JSON.parse(payloadJSON) : null; }}
    catch (e) {{ return; }}
    window.dispatchEvent(new CustomEvent(name, {{ detail: detail }}));
  }};

  // Бүрхүүлийн дотоод суваг: session cookie нь HttpOnly бөгөөд API origin-д
  // харьяалагддаг тул түүнийг зөвхөн энэ webview өөрөө хүлээж авч чадна.
  // Гэрээний хэсэг БИШ — web app үүнийг дуудахгүй.
  window.__geregeShellFetch = function (requestJSON) {{
    var req;
    try {{ req = JSON.parse(requestJSON); }}
    catch (e) {{ return; }}
    var init = {{ method: req.method || 'GET', credentials: 'include', headers: req.headers || {{}} }};
    if (req.body !== null && req.body !== undefined) {{ init.body = req.body; }}
    fetch(config.apiBase + req.path, init)
      .then(function (res) {{
        return res.text().then(function (text) {{
          return {{ id: req.id, ok: true, status: res.status, body: text }};
        }});
      }})
      .catch(function (err) {{
        return {{ id: req.id, ok: false, status: 0, body: '', error: String((err && err.message) || err) }};
      }})
      .then(function (result) {{ ipc('shell_http_result', {{ result: result }}).catch(function () {{}}); }});
  }};

  // Цэсний шошгыг сервер орчуулдаг тул бүрхүүл хэрэглэгчийн сонгосон хэлийг
  // мэдэх ёстой. Web app хэлээ localStorage-д хадгалдаг — гэрээнд хэл дамжуулах
  // method байхгүй тул бүрхүүл өөрөө уншиж авна.
  try {{
    ipc('shell_locale', {{ locale: window.localStorage.getItem('locale') || 'mn' }})
      .catch(function () {{}});
  }} catch (e) {{ /* localStorage хаалттай байж болно */ }}

  window.GeregeShell = Object.freeze({{
    version: config.version,
    platform: config.platform,
    capabilities: Object.freeze(config.capabilities.slice()),
    invoke: function (method, params) {{
      return ipc('shell_invoke', {{ method: String(method), params: params || {{}} }});
    }},
    on: function (name, handler) {{
      var listener = function (event) {{ handler(event.detail); }};
      window.addEventListener(name, listener);
      return function () {{ window.removeEventListener(name, listener); }};
    }}
  }});
}})();
"#,
        config = js_string_literal(&config_json)
    )
}

/// Бүрхүүлээс ажлын муж руу event илгээнэ.
///
/// Tauri-гийн event системийг биш, гэрээний `__geregeShellEmit`-ийг ашиглаж
/// байгаа нь санаатай: macOS-ийн лавлагаа хэрэгжилттэй яг ижил зам, бөгөөд
/// remote хуудсанд Tauri-гийн JS API байхыг шаардахгүй.
pub fn emit(window: &WebviewWindow, event: &str, payload: Value) {
    let payload_json = payload.to_string();
    let script = format!(
        "window.__geregeShellEmit && window.__geregeShellEmit({}, {});",
        js_string_literal(event),
        js_string_literal(&payload_json)
    );
    if let Err(err) = window.eval(&script) {
        eprintln!("shell: event илгээж чадсангүй ({event}): {err}");
    }
}

/// Гол цонх руу event илгээнэ. Цонх байхгүй бол чимээгүй өнгөрнө — deep link
/// нь цонх үүсэхээс өмнө ирж болно.
pub fn emit_to_main(app: &AppHandle, event: &str, payload: Value) {
    if let Some(window) = app.get_webview_window(MAIN_WINDOW) {
        emit(&window, event, payload);
    }
}

/// `shell:navigate` — зөвхөн апп доторх зам. "//host" нь протокол-харьцангуй
/// гадаад хаяг тул хасагдана.
pub fn navigate(app: &AppHandle, path: &str) {
    if !path.starts_with('/') || path.starts_with("//") {
        return;
    }
    emit_to_main(app, EVENT_NAVIGATE, json!({ "path": path }));
}

pub fn search(app: &AppHandle, query: &str) {
    emit_to_main(app, EVENT_SEARCH, json!({ "query": query }));
}

pub fn menu_refresh(app: &AppHandle) {
    emit_to_main(app, EVENT_MENU_REFRESH, Value::Null);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn literal_escapes_quotes_and_newlines() {
        assert_eq!(js_string_literal("a\"b"), "\"a\\\"b\"");
        assert_eq!(js_string_literal("a\nb"), "\"a\\nb\"");
    }

    #[test]
    fn literal_escapes_line_separators() {
        assert_eq!(js_string_literal("a\u{2028}b"), "\"a\\u2028b\"");
    }

    #[test]
    fn contract_does_not_claim_unimplemented_capabilities() {
        let declared = capabilities();
        assert!(!declared.contains(&"biometric"));
        assert!(!declared.contains(&"secure-store"));
    }
}
